pennylane/io/qasm_interpreter.py

```diff
@@ -472,6 +594,34 @@ def visit_array_literal(self, node: ArrayLiteral, context: Context):
         """
         return [self.visit(literal, context) for literal in node.values]
 
+    @visit.register(SubroutineDefinition)
+    def visit_subroutine_definition(self, node: QASMNode, context: Context):
+        """
+        Registers a subroutine definition. Maintains a namespace in the context, starts populating it with
+        its parameters.
+        Args:
+            node (QASMNode): the subroutine node.
+            context (Context): the current context.
+        """
+        context.init_subroutine_scope(node)
+
+        # register the params
+        for param in node.arguments:
+            if not isinstance(param, QuantumArgument):
+                context.scopes["subroutines"][_resolve_name(node)].vars[_resolve_name(param)] = (
+                    Variable(
+                        ty=param.__class__.__name__,
+                        val=None,
+                        size=-1,
+                        line=param.span.start_line,
+                        constant=False,
+                    )
+                )
+            else:
+                context.scopes["subroutines"][_resolve_name(node)].wires.append(
+                    _resolve_name(param)
+                )
```

Comment:
Little easier to read to control on the true case rather than the false case.
```suggestion
            if isinstance(param, QuantumArgument):
               context.scopes["subroutines"][_resolve_name(node)].wires.append(
                    _resolve_name(param)
                )
            else:
                context.scopes["subroutines"][_resolve_name(node)].vars[_resolve_name(param)] = (
                    Variable(
                        ty=param.__class__.__name__,
                        val=None,
                        size=-1,
                        line=param.span.start_line,
                        constant=False,
                    )
                )
```